### ISSUE TITLE
MPI module: slurm-launched spmd/hello.chpl prints incorrect order

### ISSUE THREAD

**Issue by ben-albrecht** [chapel-lang/chapel#5582]
### Summary of Problem

When running a Chapel program with the MPI module in SPMD mode via slurm launcher on a Cray (config details below), it appears that `MPI_Barrier` is ignored. For example, the SPMD hello test (shown below) prints the hellos out of order:

```
Hello, Chapel! This is MPI rank=2 of size=4, on locale.id=0
Hello, Chapel! This is MPI rank=1 of size=4, on locale.id=0
Hello, Chapel! This is MPI rank=0 of size=4, on locale.id=0
Hello, Chapel! This is MPI rank=3 of size=4, on locale.id=0
```

### Summary of Solution

When launching Chapel programs across multiple ranks with `CHPL_COMM=none`, it appears that buffered I/O becomes an issue, and manual flushing is required.

Follow-ups to this issues should be:

- [ ] Update `test/modules/packages/mpi/spmd/hello-chapel/hello.chpl` to be:

```chapel
use MPI;

extern proc fflush(stream);

var rank = commRank(CHPL_COMM_WORLD),
    size = commSize(CHPL_COMM_WORLD);

for irank in 0.. #size {
  if irank == rank {
    writef("Hello, Chapel! This is MPI rank=%i of size=%i, on locale.id=%i\n",rank, size, here.id);
    fflush(chpl_cstdout()); // is there a better interface for stdout-flushing in Chapel?
  }
  C_MPI.MPI_Barrier(CHPL_COMM_WORLD);
}
```

- [ ] Add a `.. tip:` to the MPI docs, mentioning how to overcome this issue.

### Steps to Reproduce

**Source Code:**

Here is the contents of the SPMD hello test:

```chapel
use MPI;

var rank = commRank(CHPL_COMM_WORLD),
    size = commSize(CHPL_COMM_WORLD);

for irank in 0.. #size {
  if irank == rank then
    writef("Hello, Chapel! This is MPI rank=%i of size=%i, on locale.id=%i\n",rank, size, here.id);
  C_MPI.MPI_Barrier(CHPL_COMM_WORLD);
}
```

**Compilation command:**
`chpl hello.chpl -o hello`

**Execution command:**
`srun --nodes=4 --ntasks=4 --ntasks-per-node=1 ./hello_real`

**Associated Future Test(s):**
`test/modules/packages/mpi/spmd/hello-chapel/hello.chpl`

### Configuration Information

- Output of `chpl --version`: `1.14.0.9316a1d` (~HEAD of master at the time of writing)
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: cray-xc
CHPL_TARGET_COMPILER: cray-prgenv-gnu *
CHPL_TARGET_ARCH: sandybridge
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none *
CHPL_TASKS: fifo *
CHPL_LAUNCHER: slurm-srun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_MAKE: gmake
CHPL_ATOMICS: intrinsics
CHPL_GMP: none *
CHPL_HWLOC: none
CHPL_REGEXP: none *
CHPL_WIDE_POINTERS: struct
CHPL_AUX_FILESYS: none

# Using Cray-MPICH 7.5.3
export MPICH_MAX_THREAD_SAFETY=multiple
```

- (For Cray systems only) Output of `module list`:
```bash
Currently Loaded Modulefiles:
  1) modules/3.2.10.6                                                         13) craype-network-aries
  2) nodestat/2.2-1.0502.60539.1.31.ari                                       14) craype-sandybridge
  3) sdb/1.1-1.0502.63652.4.25.ari                                            15) craype/2.5.10.6
  4) lustre-cray_ari_s/2.5_3.0.101_0.46.1_1.0502.8871.20.1-1.0502.21538.30.1  16) cray-mpich/7.5.3
  5) udreg/2.3.2-1.0502.10518.2.17.ari                                        17) slurm/16.05.8
  6) ugni/6.0-1.0502.10863.8.29.ari                                           18) totalview-support/1.2.0.13
  7) gni-headers/4.0-1.0502.10859.7.8.ari                                     19) totalview/2017.0.12
  8) dmapp/7.0.1-1.0502.11080.8.76.ari                                        20) cray-libsci/16.11.1
  9) xpmem/0.1-2.0502.64982.5.3.ari                                           21) pmi/5.0.11
 10) hss-llm/7.2.0                                                            22) atp/2.0.1.2
 11) Base-opts/1.0.2-1.0502.60680.2.4.ari                                     23) PrgEnv-gnu/5.2.82
 12) gcc/6.3.0                                                                24) chapel/css
```

**Comment by ben-albrecht** [comment]
@npadmana - you may also be interested in this one.

**Comment by npadmana** [comment]
@ben-albrecht -- I haven't had a chance to look at this, but here's a thought that I wanted to capture before I forgot about it. If I remember correctly, `writeln` and friends in a multi-locale environment start up a task on locale 0. I'm guessing that isn't being barrier-ed by the MPI Barrier, and so is coming out in different orders. It would be interesting to insert a Chapel barrier before the MPI Barrier to see if this fixes it.

In general, the GASNet+MPI interop prescribes that one is either in GASNet mode or MPI mode --- and that one should barrier between these.

**Comment by ben-albrecht** [comment]
Looks like I mis-diagnosed the problem... This was a case of the classic buffered I/O problem. Adding a `fflush(chpl_cstdout());` to the inner loop results in a consistently correct output.

**Comment by ben-albrecht** [comment]
Fixed in #5859
